src/ReadonlyArray.ts

```diff
@@ -0,0 +1,1678 @@
+/**
+ * @since 2.5.0
+ */
+import { Alternative1 } from './Alternative'
+import { Applicative } from './Applicative'
+import { Compactable1, Separated } from './Compactable'
+import { Either } from './Either'
+import { Eq } from './Eq'
+import { Extend1 } from './Extend'
+import { FilterableWithIndex1 } from './FilterableWithIndex'
+import { Foldable1 } from './Foldable'
+import { FoldableWithIndex1 } from './FoldableWithIndex'
+import { Predicate, Refinement, identity } from './function'
+import { FunctorWithIndex1 } from './FunctorWithIndex'
+import { HKT } from './HKT'
+import { Monad1 } from './Monad'
+import { Monoid } from './Monoid'
+import { ReadonlyNonEmptyArray } from './ReadonlyNonEmptyArray'
+import { isSome, none, Option, some } from './Option'
+import { fromCompare, getMonoid as getOrdMonoid, Ord, ordNumber } from './Ord'
+import { pipeable } from './pipeable'
+import { Show } from './Show'
+import { TraversableWithIndex1 } from './TraversableWithIndex'
+import { Unfoldable1 } from './Unfoldable'
+import { Witherable1 } from './Witherable'
+
+declare module './HKT' {
+  interface URItoKind<A> {
+    readonly ReadonlyArray: ReadonlyArray<A>
+  }
+}
+
+/**
+ * @since 2.5.0
+ */
+export const URI = 'ReadonlyArray'
+
+/**
+ * @since 2.5.0
+ */
+export type URI = typeof URI
+
+/**
+ * @since 2.5.0
+ */
+// tslint:disable-next-line: readonly-array
+export function fromArray<A>(as: Array<A>): ReadonlyArray<A> {
+  const l = as.length
+  if (l === 0) {
+    return empty
+  }
+  const ras = Array(l)
+  for (let i = 0; i < l; i++) {
+    ras[i] = as[i]
+  }
+  return ras
+}
+
+/**
+ * @since 2.5.0
+ */
+// tslint:disable-next-line: readonly-array
+export function toArray<A>(ras: ReadonlyArray<A>): Array<A> {
+  const l = ras.length
+  const as = Array(l)
+  for (let i = 0; i < l; i++) {
+    as[i] = ras[i]
+  }
+  return as
+}
+
+/**
+ * @since 2.5.0
+ */
+export function getShow<A>(S: Show<A>): Show<ReadonlyArray<A>> {
+  return {
+    show: as => `[${as.map(S.show).join(', ')}]`
+  }
+}
+
+const concat = <A>(x: ReadonlyArray<A>, y: ReadonlyArray<A>): ReadonlyArray<A> => {
```

Comment:
can we reuse `Array` module?

Reply:
same for other `Readonly*` modules

Reply:
it would bloat the bundle, I'd prefer to keep them separated

Reply:
Do you mean broken tree-shaking?

Reply:
Maybe at least reusing `ReadonlyArray` in `Array` module?

Reply:
Do you mean?

```ts
// Array.ts

import * as RA from './ReadonlyArray'

export const getShow: <A>(S: Show<A>) => Show<Array<A>> = RA.getShow

export const getMonoid: <A = never>() => Monoid<Array<A>> = RA.getMonoid as any

// etc...
```

Reply:
Yeah, exactly, however it would be nice to avoid casts, is it possible to reuse `Array` from `ReadonlyArray` since `ReadonlyArray`? That's what I meant originally.

Reply:
Some casts are unavoidable, for example `getMonoid`, `makeBy`, `range`, `replicate`, `flatten`, etc... (actually in the majority of cases)